nixops/deployment.py

```diff
@@ -1699,5 +1679,12 @@ def _load_modules_from(dir: str) -> None:
         importlib.import_module("nixops." + dir + "." + module[:-3])
 
 
+class DeploymentPlugin(Protocol):
+    def physical_spec(
+        self, d: Deployment
+    ) -> Dict[str, List[Dict[Tuple[str, ...], Any]]]:
```

Comment:
I think we should probably give this type a name, like:

```
MachineSpecs = Dict[str, List[Dict[Tuple[str, ...], Any]]]
```

or something?

Reply:
Yes, this type is incredibly awkward, it's so awkward in fact that I don't expect plugin authors to directly type these structures.

Two things I have considered instead:
- Simple dict:
``` python
{
  "networking": {"hostName": "foobar"}
}
```

- Abstract away in a class:
``` python
spec = Spec()
spec.add_option("networking.hostName", "foobar")
```